### PR TITLE
TS

### PR DESCRIPTION
**Pull request by blockarchitech**
Typescript support

### CODE REVIEW

**Comment by blockarchitech** [comment]
@cerqiest can you figure out why this doesn't build? I'm probably missing something insanely stupid.

**Comment by cerqiest** [comment]
I'll take a look. What's the error?

File: public/hosted/rs/tc.js [hunk 1]
```diff
@@ -4,3 +4,17 @@ document.addEventListener("DOMContentLoaded", function () {
     "background: #222; color: #bada55; font-size: 24px; padding: 10px;"
   );
 });
+
+// get hidden input value
+var apiKey = document.getElementById("apikey_hidden");
+document.getElementById("api_key").addEventListener("mouseover", function () {
+  let api_key = document.getElementById("api_key");
```
Comment (cerqiest) [inline public/hosted/rs/tc.js:11]: The names `apiKey` and `api_key` are a bit confusing and should probably be renamed.

File: public/stylesheets/style.css [hunk 2]
```diff
@@ -1,8 +1,73 @@
...
+
+/* Warning box (yellow) */
+
+.warning {
+	background-color: #ffffe0;
+	border: 1px solid #e6db55;
+	padding: 10px;
+	margin: 10px 0;
+}
+
```
Comment (cerqiest) [inline public/stylesheets/style.css:67]: Why are there so many blank lines?

File: src/app.ts [hunk 3]
```diff
@@ -1,14 +1,22 @@
...
-const cookieParser = require("cookie-parser");
-const logger = require("morgan");
+// app.ts
 
-const indexRouter = require("./routes/index");
-const apiRouter = require("./routes/api");
-const oauthRouter = require("./routes/oauth");
+import express from "express";
+import createError from "http-errors";
+import * as path from "path";
```
Comment (cerqiest) [inline src/app.ts:5]: Is the `* as path` nessecary?

File: src/app.ts [hunk 4]
```diff
@@ -1,14 +1,22 @@
...
+import cookieParser from "cookie-parser";
+import logger from "morgan";
+import { default as indexRouter } from "./routes/index.ts";
```
Comment (cerqiest) [inline src/app.ts:8]: Can default exports be accessed this way? I would just use `import indexRouter from ./routes/index.ts`.

File: src/app.ts [hunk 5]
```diff
@@ -1,14 +1,22 @@
...
+import { default as apiRouter } from "./routes/api.ts";
+import { default as oauthRouter } from "./routes/oauth.ts";
+import session from "express-session";
 
-const app = express();
+
+require("dotenv").config({ path: path.join(__dirname, "./.env") });
+const app: express.Application = express();
```
Comment (cerqiest) [inline src/app.ts:15]: Can the type be inferred here?

File: src/app.ts [hunk 6]
```diff
@@ -1,14 +1,22 @@
...
+import cookieParser from "cookie-parser";
+import logger from "morgan";
+import { default as indexRouter } from "./routes/index.ts";
+import { default as apiRouter } from "./routes/api.ts";
+import { default as oauthRouter } from "./routes/oauth.ts";
+import session from "express-session";
 
-const app = express();
+
+require("dotenv").config({ path: path.join(__dirname, "./.env") });
```
Comment (cerqiest) [inline src/app.ts:14]: Require isn't allowed here, no?

File: src/app.ts [hunk 7]
```diff
@@ -1,14 +1,22 @@
...
+const app: express.Application = express();
+
+/**
```
Comment (cerqiest) [inline src/app.ts:17]: Why is this comment here?

File: src/app.ts [hunk 8]
```diff
@@ -19,7 +27,11 @@ app.use(express.json());
 app.use(express.urlencoded({ extended: false }));
 app.use(cookieParser());
 app.use(express.static(path.join(__dirname, "public")));
-
+app.use(
+  session({
+    secret: process.env.SESSION_SECRET!
+  })
+);
```
Comment (cerqiest) [inline src/app.ts:34]: Whitespace should be added here.

File: src/app.ts [hunk 9]
```diff
@@ -29,7 +41,7 @@ app.use(function (req, res, next) {
 });
 
 // error handler
-app.use(function (err, req, res, next) {
+app.use(function (err: any, req: any, res: any, next: any) {
```
Comment (cerqiest) [inline src/app.ts:44]: You have the `@types/node` module. I'll see what I can do to get it to work.

**Review by zNotChill — APPROVED** [review]
what

File: src/app.ts [hunk 10]
```diff
@@ -29,7 +41,7 @@ app.use(function (req, res, next) {
...
   // set locals, only providing error in development
   res.locals.message = err.message;
```
Comment (cerqiest) [inline src/app.ts:46]: Is setting locals the best idea here? I would just pass the data directly into the template.

File: src/app.ts [hunk 11]
```diff
@@ -38,8 +50,13 @@ app.use(function (err, req, res, next) {
     res.send({ msg: err.message, meta: "https://http.cat/" + err.status });
   } else {
     res.status(err.status || 500);
+    console.log(err);
     res.send(`<img src="https://http.cat/${err.status}" alt="error" />`);
   }
 });
 
 module.exports = app;
```
Comment (cerqiest) [inline src/app.ts:58]: Is `module.exports` still valid in TS?

File: src/lib/auth.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
+// Desc: Middleware for checking if a user is authorized to access a resource
+import createError from "http-errors";
+import { db } from "./firebase.js";
+import * as Express from "express";
+
+async function check_user_key(user_id: string, key: string): Promise<boolean> {
```
Comment (cerqiest) [inline src/lib/auth.ts:6]: Can the return type be inferred?

File: src/lib/auth.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  try {
```
Comment (cerqiest) [inline src/lib/auth.ts:7]: Add some whitespace in this block.

File: src/lib/auth.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    return false;
+  } catch (err) {
+    console.error(err);
+    return false;
+  }
+}
+
+async function check_site_key(user_id: string, key: string, site_id: string): Promise<boolean> {
+  try {
```
Comment (cerqiest) [inline src/lib/auth.ts:25]: Add some whitespace in this block.

File: src/lib/auth.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+      return false;
+    }
+    return true;
+  } catch (err) {
+    console.error(err);
+    return false;
+  }
+}
+
+async function checkKeyValid(user_id: string, key: string, site_id: string | null = null): Promise<boolean> {
```
Comment (cerqiest) [inline src/lib/auth.ts:49]: Camel case for the arguments. Can the return type be inferred?

File: src/lib/auth.ts [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    const res = await check_user_key(user_id, key);
+    return res;
+  } else {
+    const res = await check_site_key(user_id, key, site_id);
+    return res;
+  }
+}
+
+
+function checker(req: Express.Request, res: Express.Response, next: Express.NextFunction): any {
```
Comment (cerqiest) [inline src/lib/auth.ts:60]: If you can use the types here, can't you use them in app.ts? Can the return type be not any?

File: src/lib/auth.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+	const headers = req.headers;
+	const apiKey = Array.isArray(headers["x-api-key"]) ? headers["x-api-key"][0] : headers["x-api-key"];
```
Comment (cerqiest) [inline src/lib/auth.ts:62]: How would headers be arrays? Do you or Express parse the headers as JSON somewhere?

File: src/lib/firebase.ts [hunk 18]
```diff
@@ -1,14 +1,14 @@
-const admin = require("firebase-admin");
-const path = require("path");
-const { v4: uuidv4 } = require("uuid");
+import admin from "firebase-admin";
+import path from "path";
+import { v4 as uuidv4 } from "uuid";
 
 const serviceAccount = require(path.join(
```
Comment (cerqiest) [inline src/lib/firebase.ts:5]: Can require be used here?

File: src/lib/zwss.ts [hunk 19]
```diff
@@ -3,11 +3,22 @@
...
 
-const yaml = require("yaml");
-const uuidv4 = require("uuid").v4;
-const { z } = require("zod");
+import yaml from "yaml";
+import { v4 as uuidv4 } from "uuid";
+import { z } from "zod";
+import { ZwssYaml, Block } from "../types/types.ts";
+
+type zwss = {
```
Comment (cerqiest) [inline src/lib/zwss.ts:11]: Does this need a type?

File: src/lib/zwss.ts [hunk 20]
```diff
@@ -55,7 +66,7 @@ zwss.validateObject = function (doc) {
  * @returns ZWSS object
  * @throws {Error} Will throw an error if the document does not match the schema
  */
-zwss.validateString = function (contents) {
+zwss.validateString = function (contents): ZwssYaml {
```
Comment (cerqiest) [inline src/lib/zwss.ts:69]: Can the return type be inferred?

File: src/lib/zwss.ts [hunk 21]
```diff
@@ -167,20 +178,20 @@ zwss.generate = function () {
...
 zwss.addBlock = function (contents, block) {
-  let doc;
+  let doc: ZwssYaml;
 
   try {
-    doc = zwss.validateString(contents);
+    doc = zwss.validateString(contents)! as ZwssYaml;
   } catch (e) {
-    throw new Error(e);
+    throw new Error(e as unknown as string);
```
Comment (cerqiest) [inline src/lib/zwss.ts:186]: Is `as unknown as string` necessary here? What does it do?

File: src/lib/zwss.ts [hunk 22]
```diff
@@ -167,20 +178,20 @@ zwss.generate = function () {
...
   }
 
-  doc.body.blocks.push(block);
+  doc.body.blocks.push(block as unknown as Block);
```
Comment (cerqiest) [inline src/lib/zwss.ts:189]: Are the `as` statements converting to unknown neccessary anywhere?

File: src/routes/index.ts [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+import { zwss } from "../lib/zwss.js";
+import { db } from "../lib/firebase.js";
+import { SiteArray } from "../types/types.js";
+
+const router: express.Router = express.Router();
+
+/* GET home page. */
+router.get("/", function (req, res, next) {
+  let no_of_sites = 0;
+  db.collection("sites")
```
Comment (cerqiest) [inline src/routes/index.ts:13]: Isn't there a way to count the number of docs in a collection?

File: src/routes/oauth.ts [hunk 24]
```diff
@@ -0,0 +1,84 @@
+// Desc: OAuth router for the app
+import * as express from "express";
+import { db } from "../lib/firebase.js";
+import { v4 as uuidv4 } from "uuid";
+import fetch from "node-fetch";
+import * as path from "path";
+import { configDotenv } from "dotenv";
+configDotenv({
```
Comment (cerqiest) [inline src/routes/oauth.ts:8]: Does this also need to be here if it's run in app.ts?

File: src/routes/oauth.ts [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+	path: path.join(__dirname, "../.env")
+})
+const router: express.Router = express.Router();
+
+router.get("/login", function (req, res, next) {
+  res.redirect(
+    `https://discord.com/api/oauth2/authorize?client_id=1137864637380034660&redirect_uri=http%3A%2F%2F127.0.0.1%3A3000%2Foauth%2Fcallback&response_type=code&scope=identify`
```
Comment (cerqiest) [inline src/routes/oauth.ts:15]: This shouldn't be hardcoded.

File: src/www.ts [hunk 26]
```diff
@@ -1,39 +1,40 @@
...
 /**
  * Module dependencies.
  */
 
-var app = require('../app');
-var debug = require('debug')('web:server');
-var http = require('http');
+import * as app from "./app.ts";
+import debug from "debug";
+debug("web:server");
```
Comment (cerqiest) [inline src/www.ts:8]: Move this below the import statements.

**Review by cerqiest — CHANGES_REQUESTED** [review]
I started losing quite a bit of my patience so I likely missed some things. See review comments. I read up on the `as unknown` thing and I now see why it is used.

File: public/hosted/rs/tc.js [hunk 27]
```diff
@@ -4,3 +4,17 @@ document.addEventListener("DOMContentLoaded", function () {
     "background: #222; color: #bada55; font-size: 24px; padding: 10px;"
   );
 });
+
+// get hidden input value
+var apiKey = document.getElementById("apikey_hidden");
+document.getElementById("api_key").addEventListener("mouseover", function () {
```
[the remark this answers is not in the document]
Comment (blockarchitech) [reply public/hosted/rs/tc.js:10]: Not needed.

File: src/app.ts (hunk 3 above)
Comment (blockarchitech) [reply src/app.ts:5]: Yes.

File: src/app.ts (hunk 4 above)
Comment (blockarchitech) [reply src/app.ts:8]: This is required as module resolution is acting strange.

File: src/app.ts (hunk 5 above)
Comment (blockarchitech) [reply src/app.ts:15]: Please elaborate.

File: src/app.ts (hunk 9 above)
Comment (blockarchitech) [reply src/app.ts:40]: 👍

File: src/app.ts (hunk 10 above)
Comment (blockarchitech) [reply src/app.ts:42]: It is for development mode only, so it is simpler this way.

File: src/app.ts [hunk 28]
```diff
@@ -38,8 +50,13 @@ app.use(function (err, req, res, next) {
     res.send({ msg: err.message, meta: "https://http.cat/" + err.status });
   } else {
     res.status(err.status || 500);
+    console.log(err);
```
[the remark this answers is not in the document]
Comment (blockarchitech) [reply src/app.ts:49]: This was temporary

File: src/app.ts (hunk 11 above)
Comment (blockarchitech) [reply src/app.ts:53]: Nope.

File: src/app.ts [hunk 29]
```diff
@@ -38,8 +50,13 @@ app.use(function (err, req, res, next) {
...
     res.send(`<img src="https://http.cat/${err.status}" alt="error" />`);
   }
 });
 
 module.exports = app;
+export function set(arg0: string, port: any) {
```
[the remark this answers is not in the document]
Comment (blockarchitech) [reply src/app.ts:54]: My mistake.

File: src/lib/auth.ts (hunk 12 above)
Comment (blockarchitech) [reply src/lib/auth.ts]: No.

File: src/lib/auth.ts (hunk 16 above)
Comment (blockarchitech) [reply src/lib/auth.ts]: Return type is any for (presumably) the same issue as app.ts.

File: src/lib/auth.ts (hunk 17 above)
Comment (blockarchitech) [reply src/lib/auth.ts]: Express sends me the headers as an array.

File: src/lib/firebase.ts (hunk 18 above)
Comment (blockarchitech) [reply src/lib/firebase.ts:5]: No.

File: src/lib/zwss.ts (hunk 19 above)
Comment (blockarchitech) [reply src/lib/zwss.ts:11]: It makes me feel better :)

File: src/lib/zwss.ts (hunk 20 above)
Comment (blockarchitech) [reply src/lib/zwss.ts:68]: It cannot.

File: src/lib/zwss.ts (hunk 21 above)
Comment (blockarchitech) [reply src/lib/zwss.ts:185]: Some errors can be blank.

File: src/lib/zwss.ts (hunk 22 above)
Comment (blockarchitech) [reply src/lib/zwss.ts:188]: No idea! I just did this and the red line went away.

File: src/routes/index.ts (hunk 23 above)
Comment (blockarchitech) [reply src/routes/index.ts:13]: You'd think.

File: src/routes/oauth.ts (hunk 24 above)
Comment (blockarchitech) [reply src/routes/oauth.ts:8]: Yes.

File: src/routes/oauth.ts (hunk 25 above)
Comment (blockarchitech) [reply src/routes/oauth.ts:15]: Will be changed - discord auth has a bug where `encodeURIIbject` doesn't work in OAuth URI's. Discord is aware of this bug.

File: src/www.ts (hunk 26 above)
Comment (blockarchitech) [reply src/www.ts:7]: I cannot. Some modules need imported before others.

**Comment by blockarchitech** [comment]
@cerqiest review

**Review by cerqiest — APPROVED** [review]
I do not have the time for this

**Pull request merged**